[PATCH] TOMOYO 2.4 backport patch for Linux 2.6.39

Apply this patch after overwriting security/tomoyo/ directory by doing

 $ wget -O linux-3.1-rc7.tar.gz https://github.com/torvalds/linux/tarball/v3.1-rc7
 $ tar -zxf linux-3.1-rc7.tar.gz --strip 1 torvalds-linux-4cf670b/security/tomoyo/

.

Signed-off-by: Tetsuo Handa <penguin-kernel@I-love.SAKURA.ne.jp>
---
 security/tomoyo/domain.c        |    4 +++-
 security/tomoyo/securityfs_if.c |    1 +
 security/tomoyo/util.c          |    3 ++-
 3 files changed, 6 insertions(+), 2 deletions(-)

--- linux-2.6.39.4.orig/security/tomoyo/domain.c
+++ linux-2.6.39.4/security/tomoyo/domain.c
@@ -501,7 +501,8 @@ struct tomoyo_domain_info *tomoyo_assign
 			 * that domain. Do not perform domain transition if
 			 * profile for that domain is not yet created.
 			 */
-			if (!entry->ns->profile_ptr[entry->profile])
+			if (tomoyo_policy_loaded &&
+			    !entry->ns->profile_ptr[entry->profile])
 				return NULL;
 		}
 		return entry;
@@ -557,6 +558,7 @@ out:
 			tomoyo_write_log(&r, "use_profile %u\n",
 					 entry->profile);
 			tomoyo_write_log(&r, "use_group %u\n", entry->group);
+			tomoyo_update_stat(TOMOYO_STAT_POLICY_UPDATES);
 		}
 	}
 	return entry;
--- linux-2.6.39.4.orig/security/tomoyo/securityfs_if.c
+++ linux-2.6.39.4/security/tomoyo/securityfs_if.c
@@ -147,6 +147,7 @@ static int __init tomoyo_initerface_init
 			    TOMOYO_MANAGER);
 	tomoyo_create_entry("version",          0400, tomoyo_dir,
 			    TOMOYO_VERSION);
+	tomoyo_load_builtin_policy();
 	return 0;
 }
 
--- linux-2.6.39.4.orig/security/tomoyo/util.c
+++ linux-2.6.39.4/security/tomoyo/util.c
@@ -925,7 +925,8 @@ int tomoyo_get_mode(const struct tomoyo_
 		return TOMOYO_CONFIG_DISABLED;
 	mode = tomoyo_profile(ns, profile)->config[index];
 	if (mode == TOMOYO_CONFIG_USE_DEFAULT)
-		mode = tomoyo_profile(ns, profile)->config[category];
+		mode = tomoyo_profile(ns, profile)->config
+			[category + TOMOYO_MAX_MAC_INDEX];
 	if (mode == TOMOYO_CONFIG_USE_DEFAULT)
 		mode = tomoyo_profile(ns, profile)->default_config;
 	return mode & 3;
